File: src/main/gemini/usage.ts
/**
 * Provider boundary for token/context usage.
 *
 * Everything Gemini-specific about usage reporting lives here: the standardized
 * ACP `PromptResponse.usage` field, Gemini CLI 0.56's proprietary
 * `_meta.quota.token_count` payload (snake_case) and the ACP `usage_update`
 * notification. Nothing outside this module may know about `_meta` or
 * snake_case keys, and no raw provider object is ever forwarded.
 */

export interface TokenCounters {
  readonly input: number | null;
  readonly output: number | null;
  readonly total: number | null;
  readonly thought: number | null;
  readonly cachedRead: number | null;
  readonly cachedWrite: number | null;
  readonly tool: number | null;
  readonly totalKind: "provider" | "derived_input_plus_output" | null;
}

export interface ModelTokenUsage {
  readonly model: string;
  readonly input: number;
  readonly output: number;
}

export interface UsageCost {
  readonly amount: number;
  readonly currency: string;
}

/**
 * Token counters observed for one prompt response.
 *
 * `scope` is deliberately explicit: the SDK 1.3.0 type comments describe
 * `PromptResponse.usage` as cumulative for the session, while Gemini's
 * `_meta.quota` counters are reset at the start of every `Session.prompt()`.
 * Guessing the level of validity is exactly the bug this module prevents.
 */
export interface UsageTokenObservation {
  readonly kind: "tokens";
  readonly scope: "turn" | "session_cumulative";
  readonly source: "acp_prompt_usage" | "gemini_meta_quota";
  readonly tokens: TokenCounters;
  readonly byModel: readonly ModelTokenUsage[];
}

/** Context-window occupancy. Never a token consumption value. */
export interface UsageContextObservation {
  readonly kind: "context";
  readonly source: "acp_usage_update";
  readonly used: number;
  readonly size: number;
  readonly cost: UsageCost | null;
}

export type UsageObservation = UsageTokenObservation | UsageContextObservation;

export type UsageDiagnostic = (message: string) => void;

export const EMPTY_TOKEN_COUNTERS: TokenCounters = {
  input: null,
  output: null,
  total: null,
  thought: null,
  cachedRead: null,
  cachedWrite: null,
  tool: null,
  totalKind: null,
};

/**
 * Reads the token counters of a finished prompt turn.
 *
 * Priority follows the fallback table in token-usage.md: the standardized
 * `usage` field wins, Gemini's `_meta.quota` is the fallback. When both exist
 * the standardized field defines the session state while `_meta.quota` may
 * still contribute the per-model split, so nothing is counted twice.
 */
export function parsePromptUsage(
  response: unknown,
  diagnose?: UsageDiagnostic,
): UsageTokenObservation | null {
  const record = asRecord(response);
  if (!record) return null;

  const byModel = parseModelUsage(record, diagnose);
  const standard = parseStandardUsage(record.usage, diagnose);
  if (standard) {
    return {
      kind: "tokens",
      scope: "session_cumulative",
      source: "acp_prompt_usage",
      tokens: standard,
      byModel,
    };
  }

  const quota = parseGeminiQuota(record, diagnose);
  if (quota) {
    return {
      kind: "tokens",
      scope: "turn",
      source: "gemini_meta_quota",
      tokens: quota,
      byModel,
    };
  }
  return null;
}

/** Reads an ACP `usage_update` notification body. */
export function parseUsageUpdate(
  update: unknown,
  diagnose?: UsageDiagnostic,
): UsageContextObservation | null {
  const record = asRecord(update);
  if (!record) return null;

  const used = safeCount(record.used);
  const size = safeCount(record.size);
  if (used === null || size === null || size <= 0) {
    diagnose?.("usage_update dropped: used/size are not usable token counts");
    return null;
  }
  return {
    kind: "context",
    source: "acp_usage_update",
    used,
    size,
    cost: parseCost(record.cost, diagnose),
  };
}

function parseStandardUsage(
  value: unknown,
  diagnose?: UsageDiagnostic,
): TokenCounters | null {
  const usage = asRecord(value);
  if (!usage) return null;

  const input = safeCount(usage.inputTokens);
  const output = safeCount(usage.outputTokens);
  const providerTotal = safeCount(usage.totalTokens);
  if (input === null && output === null && providerTotal === null) {
    if (value !== undefined && value !== null) {
      diagnose?.("PromptResponse.usage contained no usable token counts");
    }
    return null;
  }

  return buildCounters({
    input,
    output,
    providerTotal,
    thought: safeCount(usage.thoughtTokens),
    cachedRead: safeCount(usage.cachedReadTokens),
    cachedWrite: safeCount(usage.cachedWriteTokens),
    tool: safeCount(usage.toolTokens),
  });
}

function parseGeminiQuota(
  response: Record<string, unknown>,
  diagnose?: UsageDiagnostic,
): TokenCounters | null {
  const quota = asRecord(asRecord(response._meta)?.quota);
  if (!quota) return null;
  const counts = asRecord(quota.token_count);
  if (!counts) {
    diagnose?.("_meta.quota did not contain a token_count object");
    return null;
  }

  const input = safeCount(counts.input_tokens);
  const output = safeCount(counts.output_tokens);
  const providerTotal = safeCount(counts.total_tokens);
  if (input === null && output === null && providerTotal === null) {
    diagnose?.("_meta.quota.token_count contained no usable token counts");
    return null;
  }

  return buildCounters({
    input,
    output,
    providerTotal,
    thought: safeCount(counts.thoughts_token_count),
    cachedRead: safeCount(counts.cached_content_token_count),
    cachedWrite: null,
    tool: safeCount(counts.tool_use_prompt_token_count),
  });
}

function parseModelUsage(
  response: Record<string, unknown>,
  diagnose?: UsageDiagnostic,
): ModelTokenUsage[] {
  const quota = asRecord(asRecord(response._meta)?.quota);
  const entries = quota?.model_usage;
  if (!Array.isArray(entries)) return [];

  const result: ModelTokenUsage[] = [];
  for (const entry of entries.slice(0, 50)) {
    const record = asRecord(entry);
    const model = typeof record?.model === "string" ? record.model.trim() : "";
    const counts = asRecord(record?.token_count);
    if (!model || !counts) {
      diagnose?.("_meta.quota.model_usage entry ignored: missing model or token_count");
      continue;
    }
    const input = safeCount(counts.input_tokens) ?? 0;
    const output = safeCount(counts.output_tokens) ?? 0;
    result.push({ model: model.slice(0, 200), input, output });
  }
  return result;
}

function parseCost(value: unknown, diagnose?: UsageDiagnostic): UsageCost | null {
  const cost = asRecord(value);
  if (!cost) return null;
  const amount = cost.amount;
  const currency = cost.currency;
  if (
    typeof amount !== "number" ||
    !Number.isFinite(amount) ||
    amount < 0 ||
    typeof currency !== "string" ||
    !/^[A-Za-z]{3}$/.test(currency.trim())
  ) {
    diagnose?.("usage_update.cost dropped: amount or currency is not usable");
    return null;
  }
  return { amount, currency: currency.trim().toUpperCase() };
}

function buildCounters(input: {
  input: number | null;
  output: number | null;
  providerTotal: number | null;
  thought: number | null;
  cachedRead: number | null;
  cachedWrite: number | null;
  tool: number | null;
}): TokenCounters {
  const derived =
    input.input !== null && input.output !== null ? input.input + input.output : null;
  const total = input.providerTotal ?? derived;
  const totalKind =
    input.providerTotal !== null
      ? ("provider" as const)
      : derived !== null
        ? ("derived_input_plus_output" as const)
        : null;

  return {
    input: input.input,
    output: input.output,
    total: total !== null && Number.isSafeInteger(total) ? total : null,
    thought: input.thought,
    cachedRead: input.cachedRead,
    cachedWrite: input.cachedWrite,
    tool: input.tool,
    totalKind: total !== null && Number.isSafeInteger(total) ? totalKind : null,
  };
}

/** Adds two counter sets. Unknown values stay unknown instead of becoming zero. */
export function addTokenCounters(
  left: TokenCounters,
  right: TokenCounters,
): TokenCounters {
  const input = addCounts(left.input, right.input);
  const output = addCounts(left.output, right.output);
  const total = addCounts(left.total, right.total);
  const derivedOnly =
    left.totalKind !== "provider" && right.totalKind !== "provider";
  return {
    input,
    output,
    total,
    thought: addCounts(left.thought, right.thought),
    cachedRead: addCounts(left.cachedRead, right.cachedRead),
    cachedWrite: addCounts(left.cachedWrite, right.cachedWrite),
    tool: addCounts(left.tool, right.tool),
    totalKind:
      total === null ? null : derivedOnly ? "derived_input_plus_output" : "provider",
  };
}

function addCounts(left: number | null, right: number | null): number | null {
  if (left === null) return right;
  if (right === null) return left;
  const sum = left + right;
  return Number.isSafeInteger(sum) ? sum : null;
}

/**
 * Accepts only non-negative safe integers. Negative, non-finite, fractional and
 * unsafe values are dropped instead of being rounded into a plausible lie.
 */
export function safeCount(value: unknown): number | null {
  if (typeof value !== "number") return null;
  if (!Number.isSafeInteger(value) || value < 0) return null;
  return value;
}

function asRecord(value: unknown): Record<string, unknown> | null {
  if (!value || typeof value !== "object" || Array.isArray(value)) return null;
  return value as Record<string, unknown>;
}

File: src/acp/usage.rs
//! Defensive normalization of ACP/Gemini usage payloads.
//!
//! ACP's standard `usage` object is cumulative according to the SDK contract,
//! while Gemini CLI 0.56 exposes per-prompt counters under `_meta.quota`.
//! Keeping the scope explicit is what prevents a fallback counter from being
//! added twice after a reconnect.

use serde::{Deserialize, Serialize};
use serde_json::Value;

#[derive(Clone, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct TokenCounters {
    pub input: Option<u64>,
    pub output: Option<u64>,
    pub total: Option<u64>,
    pub thought: Option<u64>,
    pub cached_read: Option<u64>,
    pub cached_write: Option<u64>,
    pub tool: Option<u64>,
    pub total_kind: Option<TotalKind>,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum TotalKind {
    Provider,
    DerivedInputPlusOutput,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct ModelTokenUsage {
    pub model: String,
    pub input: u64,
    pub output: u64,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum UsageScope {
    Turn,
    SessionCumulative,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum TokenUsageSource {
    AcpPromptUsage,
    GeminiMetaQuota,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct UsageTokenObservation {
    pub kind: String,
    pub scope: UsageScope,
    pub source: TokenUsageSource,
    pub tokens: TokenCounters,
    pub by_model: Vec<ModelTokenUsage>,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct UsageCost {
    pub amount: f64,
    pub currency: String,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct UsageContextObservation {
    pub kind: String,
    pub source: String,
    pub used: u64,
    pub size: u64,
    pub cost: Option<UsageCost>,
}

pub fn parse_prompt_usage(response: &Value) -> Option<UsageTokenObservation> {
    let object = response.as_object()?;
    let by_model = parse_model_usage(object.get("_meta"));
    if let Some(tokens) = parse_standard_usage(object.get("usage")) {
        return Some(UsageTokenObservation {
            kind: "tokens".into(),
            scope: UsageScope::SessionCumulative,
            source: TokenUsageSource::AcpPromptUsage,
            tokens,
            by_model,
        });
    }
    parse_gemini_quota(object.get("_meta")).map(|tokens| UsageTokenObservation {
        kind: "tokens".into(),
        scope: UsageScope::Turn,
        source: TokenUsageSource::GeminiMetaQuota,
        tokens,
        by_model,
    })
}

pub fn parse_usage_update(update: &Value) -> Option<UsageContextObservation> {
    let object = update.as_object()?;
    let used = safe_count(object.get("used"))?;
    let size = safe_count(object.get("size"))?;
    if size == 0 {
        return None;
    }
    Some(UsageContextObservation {
        kind: "context".into(),
        source: "acp_usage_update".into(),
        used,
        size,
        cost: parse_cost(object.get("cost")),
    })
}

pub fn add_token_counters(left: &TokenCounters, right: &TokenCounters) -> TokenCounters {
    let input = add_count(left.input, right.input);
    let output = add_count(left.output, right.output);
    let total = add_count(left.total, right.total);
    let total_kind = total.map(|_| {
        if left.total_kind == Some(TotalKind::Provider)
            || right.total_kind == Some(TotalKind::Provider)
        {
            TotalKind::Provider
        } else {
            TotalKind::DerivedInputPlusOutput
        }
    });
    TokenCounters {
        input,
        output,
        total,
        thought: add_count(left.thought, right.thought),
        cached_read: add_count(left.cached_read, right.cached_read),
        cached_write: add_count(left.cached_write, right.cached_write),
        tool: add_count(left.tool, right.tool),
        total_kind,
    }
}

pub fn safe_count(value: Option<&Value>) -> Option<u64> {
    let value = value?.as_u64()?;
    // SQLite's INTEGER and JavaScript's safe integers are both bounded below
    // this. Refuse values that cannot be represented by either side.
    (value <= 9_007_199_254_740_991).then_some(value)
}

fn parse_standard_usage(value: Option<&Value>) -> Option<TokenCounters> {
    let value = value?.as_object()?;
    let input = safe_count(value.get("inputTokens"));
    let output = safe_count(value.get("outputTokens"));
    let provider_total = safe_count(value.get("totalTokens"));
    if input.is_none() && output.is_none() && provider_total.is_none() {
        return None;
    }
    Some(build_counters(
        input,
        output,
        provider_total,
        safe_count(value.get("thoughtTokens")),
        safe_count(value.get("cachedReadTokens")),
        safe_count(value.get("cachedWriteTokens")),
        safe_count(value.get("toolTokens")),
    ))
}

fn parse_gemini_quota(meta: Option<&Value>) -> Option<TokenCounters> {
    let counts = meta?.get("quota")?.get("token_count")?.as_object()?;
    let input = safe_count(counts.get("input_tokens"));
    let output = safe_count(counts.get("output_tokens"));
    let provider_total = safe_count(counts.get("total_tokens"));
    if input.is_none() && output.is_none() && provider_total.is_none() {
        return None;
    }
    Some(build_counters(
        input,
        output,
        provider_total,
        safe_count(counts.get("thoughts_token_count")),
        safe_count(counts.get("cached_content_token_count")),
        None,
        safe_count(counts.get("tool_use_prompt_token_count")),
    ))
}

fn parse_model_usage(meta: Option<&Value>) -> Vec<ModelTokenUsage> {
    let Some(entries) = meta
        .and_then(|value| value.get("quota"))
        .and_then(|value| value.get("model_usage"))
        .and_then(Value::as_array)
    else {
        return Vec::new();
    };
    entries
        .iter()
        .take(50)
        .filter_map(|entry| {
            let model = entry.get("model")?.as_str()?.trim();
            if model.is_empty() || model.chars().count() > 200 {
                return None;
            }
            let counts = entry.get("token_count")?;
            Some(ModelTokenUsage {
                model: model.to_owned(),
                input: safe_count(counts.get("input_tokens")).unwrap_or(0),
                output: safe_count(counts.get("output_tokens")).unwrap_or(0),
            })
        })
        .collect()
}

fn parse_cost(value: Option<&Value>) -> Option<UsageCost> {
    let value = value?.as_object()?;
    let amount = value.get("amount")?.as_f64()?;
    let currency = value.get("currency")?.as_str()?.trim().to_uppercase();
    if !amount.is_finite()
        || amount < 0.0
        || currency.len() != 3
        || !currency.bytes().all(|byte| byte.is_ascii_uppercase())
    {
        return None;
    }
    Some(UsageCost { amount, currency })
}

fn build_counters(
    input: Option<u64>,
    output: Option<u64>,
    provider_total: Option<u64>,
    thought: Option<u64>,
    cached_read: Option<u64>,
    cached_write: Option<u64>,
    tool: Option<u64>,
) -> TokenCounters {
    let derived = input
        .zip(output)
        .and_then(|(left, right)| left.checked_add(right));
    let total = provider_total.or(derived);
    TokenCounters {
        input,
        output,
        total,
        thought,
        cached_read,
        cached_write,
        tool,
        total_kind: total.map(|_| {
            if provider_total.is_some() {
                TotalKind::Provider
            } else {
                TotalKind::DerivedInputPlusOutput
            }
        }),
    }
}

fn add_count(left: Option<u64>, right: Option<u64>) -> Option<u64> {
    match (left, right) {
        (Some(left), Some(right)) => left.checked_add(right),
        (Some(value), None) | (None, Some(value)) => Some(value),
        (None, None) => None,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn standard_usage_wins_over_gemini_fallback() {
        let result = parse_prompt_usage(&json!({
            "usage": { "inputTokens": 10, "outputTokens": 20, "totalTokens": 30 },
            "_meta": { "quota": { "token_count": { "input_tokens": 1, "output_tokens": 2 } } }
        }))
        .unwrap();
        assert_eq!(result.source, TokenUsageSource::AcpPromptUsage);
        assert_eq!(result.scope, UsageScope::SessionCumulative);
        assert_eq!(result.tokens.total, Some(30));
    }

    #[test]
    fn parses_gemini_turn_counters_and_model_split() {
        let result = parse_prompt_usage(&json!({
            "_meta": { "quota": { "token_count": { "input_tokens": 1234, "output_tokens": 567 }, "model_usage": [{ "model": "pro", "token_count": { "input_tokens": 1200, "output_tokens": 500 } }] } }
        })).unwrap();
        assert_eq!(result.source, TokenUsageSource::GeminiMetaQuota);
        assert_eq!(
            result.tokens.total_kind,
            Some(TotalKind::DerivedInputPlusOutput)
        );
        assert_eq!(result.by_model[0].model, "pro");
    }

    #[test]
    fn rejects_invalid_context_and_keeps_cost_separate() {
        assert!(parse_usage_update(&json!({ "used": 10, "size": 0 })).is_none());
        let context = parse_usage_update(
            &json!({ "used": 10, "size": 100, "cost": { "amount": 0.1, "currency": "usd" } }),
        )
        .unwrap();
        assert_eq!(context.cost.unwrap().currency, "USD");
    }
}
